src/networking.c

```diff
@@ -3207,37 +3222,64 @@ void flushSlavesOutputBuffers(void) {
  * time left for the previous duration. However if the duration is smaller
  * than the time left for the previous pause, no change is made to the
  * left duration. */
-void pauseClients(mstime_t end) {
-    if (!server.clients_paused || end > server.clients_pause_end_time)
-        server.clients_pause_end_time = end;
-    server.clients_paused = 1;
+void pauseClients(mstime_t end, int type) {
+    mstime_t *current_end;
+
+    if (type == CLIENT_PAUSE_WRITE) {
+        current_end = &server.client_pause_write_end_time;
+    } else if (type == CLIENT_PAUSE_ALL) {
+        current_end = &server.client_pause_end_time;
+    } else {
+        serverPanic("Pause called with invalid type.");
+    }
+
+    if (!(server.client_pause_flags & type) || end > *current_end) {
+        *current_end = end;
+    }
+
+    server.client_pause_flags |= type;
 }
 
-/* Return non-zero if clients are currently paused. As a side effect the
- * function checks if the pause time was reached and clear it. */
-int clientsArePaused(void) {
-    if (server.clients_paused &&
-        server.clients_pause_end_time < server.mstime)
+/* Unpause clients if either the time has elapsed or clients
+ * need to be forced to be unpaused. */
+void unpauseClients(int force) {
```

Comment:
I think the problem is that it has the side effect of unpausing clients. I tried decomposing it a little bit. It's a little more verbose, but hopefully more readable.

Reply:
Yes, it was worse before (looking as a gettter or checker function and now it's better) since it's clear it takes action, but it doesn't always take action, and that's still a bit misleading